independent-projects/bootstrap/runner/src/test/java/io/quarkus/bootstrap/runner/RunnerClassLoaderTest.java

```diff
@@ -0,0 +1,94 @@
+package io.quarkus.bootstrap.runner;
+
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.junit.jupiter.api.Test;
+
+public class RunnerClassLoaderTest {
+
+    static {
+        System.setProperty("jdk.tracePinnedThreads", "full");
+    }
+
+    @Test
+    public void test() throws Exception {
+        Map<String, ClassLoadingResource[]> resourceDirectoryMap = new HashMap<>();
+        resourceDirectoryMap.put("org/mario",
+                new ClassLoadingResource[] { createApacheTextJarResource("a"), createMyProjectJarResource() });
+
+        // These jars are simply used to fill the RunnerClassLoader's jars cache
+        resourceDirectoryMap.put("org/apache/commons/text", new ClassLoadingResource[] {
+                createApacheExecJarResource("b"), createApacheExecJarResource("c"),
+                createApacheExecJarResource("d"), createApacheTextJarResource("a") });
+
+        RunnerClassLoader runnerClassLoader = new RunnerClassLoader(ClassLoader.getSystemClassLoader(), resourceDirectoryMap,
+                Collections.emptySet(), Collections.emptySet(),
+                Collections.emptyList(), Collections.emptyMap());
+
+        // Put the RunnerClassLoader in a postBootPhase thus enabling the jars cache
+        runnerClassLoader.resetInternalCaches();
+
+        // Fills the RunnerClassLoader's jars cache
+        runnerClassLoader.loadClass("org.apache.commons.text.AlphabetConverter");
+
+        Runnable runnable = null;
+
+        try {
+            Class<?> runnableClazz = runnerClassLoader.loadClass("org.mario.MyRunnable");
+            runnable = (Runnable) runnableClazz.getConstructor().newInstance();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        ExecutorService virtualExecutor = createVirtualThreadExecutor();
+        try {
+            for (int i = 0; i < 2; i++) {
+                virtualExecutor.submit(runnable);
```

Comment:
The code of that `MyRunnable` class that I packaged in the `myproject-1.0.jar` does the following:

```
package org.mario;

public class MyRunnable implements Runnable {

    private String className;

    @Override
    public void run() {
        className = MyPojo.class.getCanonicalName();
    }
}

```

In essence it triggers the `defineClass` (which is a native method) of the MyPojo class, also present in the same jar, through the same classloader, the instance of `RunnerClassLoader` created in the unit test, through which also the `MyRunnable` class has been loaded.

Another way to obtain the same effect is passing the `RunnerClassLoader` instance to the `Runnable` and in the run method loading the `MyPojo` class using a `Class::forName` like in

  `Class.forName("org.mario.MyPojo", true, runnerClassLoader);`

Conversely note that doing 

  `runnerClassLoader.loadClass("org.mario.MyPojo");`

doesn't have the same effect because it bypasses the native `defineClass`, thus not triggering the pinning. I'm mentioning this mostly because it made me wasting a couple of days of works :( and I discovered this mistake only thanks to @franz1981's help.